File: python/eval/model_load_speed.py
#!/usr/bin/env python3
# Copyright 2026 Jeremy Lilley (jeremy@jlilley.net)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Benchmark model loading time and memory consumption.

Usage:
    python3 model_load_speed.py --languages efigs
    python3 model_load_speed.py --languages europe_west_common --iterations 5
    python3 model_load_speed.py --languages all
"""

import argparse
import resource
import sys
import time
import tracemalloc

from langidentify import Language, Model
from langidentify.model_loader import clear_cache


def _load_model(args_model, languages):
  if args_model == "lite":
    return Model.load_lite(languages)
  elif args_model == "full":
    return Model.load_full(languages)
  else:
    return Model.load(languages)


def main():
  parser = argparse.ArgumentParser(
    description="Benchmark model loading time and memory consumption.")
  parser.add_argument(
    "--languages", required=True,
    help="Comma-separated language codes or group aliases (e.g. efigs, europe_west_common)")
  parser.add_argument(
    "--model", choices=["lite", "full", "best"], default="best",
    help="Model variant: lite, full, or best (auto-detect, default)")
  parser.add_argument(
    "--iterations", type=int, default=3,
    help="Number of loading iterations to average (default: 3)")
  args = parser.parse_args()

  languages = Language.from_comma_separated(args.languages)
  if not languages:
    print("No languages matched.", file=sys.stderr)
    sys.exit(1)

  print(f"Model:      {args.model}")
  print(f"Languages:  {len(languages)} configured")
  print(f"Iterations: {args.iterations}")
  print()

  # Time iterations without tracemalloc (which adds significant overhead).
  times = []
  for i in range(args.iterations):
    clear_cache()

    t0 = time.monotonic()
    model = _load_model(args.model, languages)
    elapsed = time.monotonic() - t0
    times.append(elapsed)

    rss_raw = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    if sys.platform == "darwin":
      rss_mb = rss_raw / (1024 * 1024)
    else:
      rss_mb = rss_raw / 1024

    print(f"  Iteration {i + 1}: {elapsed:.3f}s  RSS: {rss_mb:.1f} MB")

  avg = sum(times) / len(times)
  best = min(times)
  worst = max(times)

  print()
  print(f"Load time:  avg={avg:.3f}s  best={best:.3f}s  worst={worst:.3f}s")

  # Run one more load with tracemalloc to measure Python memory allocation.
  # This is done separately because tracemalloc adds ~8-10x overhead.
  print()
  print("Running once more with tracemalloc for accurate memory measurement "
        "(slower, RSS is not reclaimed between loads)...")
  clear_cache()
  tracemalloc.start()
  model = _load_model(args.model, languages)
  current, peak = tracemalloc.get_traced_memory()
  tracemalloc.stop()

  print(f"Memory:     {current / 1024 / 1024:.1f} MB current, "
        f"{peak / 1024 / 1024:.1f} MB peak (tracemalloc)")

  # Count entries across tables.
  total_entries = 0
  for n in range(1, model.max_ngram + 1):
    table = model.get_table(n)
    total_entries += len(table)
  tw_entries = len(model._topwords_table)

  print(f"Tables:     {total_entries:,} ngram entries, "
        f"{tw_entries:,} topword entries")


if __name__ == "__main__":
  main()

File: rust/eval/src/model_load_speed.rs
// Copyright 2026 Jeremy Lilley
// Licensed under the Apache License, Version 2.0

//! Benchmark model loading time and memory consumption.
//!
//! Usage:
//!     cargo run --release --bin model-load-speed -- --languages efigs
//!     cargo run --release --bin model-load-speed -- --languages europe_west_common --iterations 5

use langidentify::{Language, Model};
use langidentify_eval::common::ModelVariant;
use std::alloc::{GlobalAlloc, Layout, System};
use std::process;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::time::Instant;

// ========================================================================
// Counting allocator — tracks current and peak heap usage.
// ========================================================================

struct CountingAllocator;

static ALLOCATED: AtomicUsize = AtomicUsize::new(0);
static PEAK: AtomicUsize = AtomicUsize::new(0);

unsafe impl GlobalAlloc for CountingAllocator {
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        let ptr = unsafe { System.alloc(layout) };
        if !ptr.is_null() {
            let current = ALLOCATED.fetch_add(layout.size(), Ordering::Relaxed) + layout.size();
            // Update peak (relaxed CAS loop).
            let mut peak = PEAK.load(Ordering::Relaxed);
            while current > peak {
                match PEAK.compare_exchange_weak(peak, current, Ordering::Relaxed, Ordering::Relaxed)
                {
                    Ok(_) => break,
                    Err(actual) => peak = actual,
                }
            }
        }
        ptr
    }

    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
        unsafe { System.dealloc(ptr, layout) };
        ALLOCATED.fetch_sub(layout.size(), Ordering::Relaxed);
    }
}

#[global_allocator]
static GLOBAL: CountingAllocator = CountingAllocator;

fn reset_peak() {
    PEAK.store(ALLOCATED.load(Ordering::Relaxed), Ordering::Relaxed);
}

fn current_allocated_mb() -> f64 {
    ALLOCATED.load(Ordering::Relaxed) as f64 / (1024.0 * 1024.0)
}

fn peak_allocated_mb() -> f64 {
    PEAK.load(Ordering::Relaxed) as f64 / (1024.0 * 1024.0)
}

// ========================================================================

fn load_model_raw(variant: ModelVariant, languages: &[Language]) -> Model {
    match variant {
        ModelVariant::Lite => Model::load_embedded(
            langidentify_models_lite::resolve,
            languages,
            -12.0,
            -12.0,
            0.0,
        )
        .unwrap(),
        ModelVariant::Full => Model::load_embedded(
            langidentify_models_full::resolve,
            languages,
            -15.0,
            -15.0,
            0.0,
        )
        .unwrap(),
        ModelVariant::Best => {
            if langidentify_models_full::exists("skipwords.txt") {
                Model::load_embedded(
                    langidentify_models_full::resolve,
                    languages,
                    -15.0,
                    -15.0,
                    0.0,
                )
                .unwrap()
            } else {
                Model::load_embedded(
                    langidentify_models_lite::resolve,
                    languages,
                    -12.0,
                    -12.0,
                    0.0,
                )
                .unwrap()
            }
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let (languages_str, iterations, variant) = parse_args(&args);

    let languages = match Language::from_comma_separated(&languages_str) {
        Ok(langs) => langs,
        Err(e) => {
            eprintln!("Invalid languages: {}", e);
            process::exit(1);
        }
    };

    let variant_name = match variant {
        ModelVariant::Lite => "lite",
        ModelVariant::Full => "full",
        ModelVariant::Best => "best",
    };

    println!("Model:      {}", variant_name);
    println!("Languages:  {} configured", languages.len());
    println!("Iterations: {}", iterations);
    println!();

    let mut times = Vec::with_capacity(iterations);

    // Time iterations — drop the model between iterations to measure load cost.
    for i in 0..iterations {
        let t0 = Instant::now();
        let m = load_model_raw(variant, &languages);
        let elapsed = t0.elapsed().as_secs_f64();
        times.push(elapsed);
        println!("  Iteration {}: {:.3}s", i + 1, elapsed);
        drop(m);
    }

    let avg: f64 = times.iter().sum::<f64>() / times.len() as f64;
    let best = times.iter().cloned().fold(f64::INFINITY, f64::min);
    let worst = times.iter().cloned().fold(0.0f64, f64::max);

    println!();
    println!(
        "Load time:  avg={:.3}s  best={:.3}s  worst={:.3}s",
        avg, best, worst,
    );

    // Memory measurement — load once more with tracking.
    let before = current_allocated_mb();
    reset_peak();
    let model = load_model_raw(variant, &languages);
    let after = current_allocated_mb();
    let peak = peak_allocated_mb();

    println!(
        "Memory:     {:.1} MB resident, {:.1} MB peak during load",
        after - before,
        peak - before,
    );

    // Report table sizes.
    let mut total_entries = 0usize;
    for n in 1..=model.max_ngram() {
        total_entries += model.table(n).size();
    }
    let tw_entries = model.topwords_table().size();
    println!(
        "Tables:     {} ngram entries, {} topword entries",
        total_entries, tw_entries,
    );
}

fn parse_args(args: &[String]) -> (String, usize, ModelVariant) {
    let mut languages = None;
    let mut iterations = 3usize;
    let mut variant = ModelVariant::Best;
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--languages" => {
                i += 1;
                languages = Some(args[i].clone());
            }
            "--iterations" => {
                i += 1;
                iterations = args[i].parse().unwrap_or(3);
            }
            "--model" => {
                i += 1;
                variant = ModelVariant::from_str(&args[i]);
            }
            "--help" | "-h" => {
                eprintln!(
                    "Usage: model-load-speed --languages <langs> \
                     [--iterations <n>] [--model lite|full|best]"
                );
                process::exit(0);
            }
            _ => {
                eprintln!("Unknown argument: {}", args[i]);
                process::exit(1);
            }
        }
        i += 1;
    }
    let languages = languages.unwrap_or_else(|| {
        eprintln!("--languages is required");
        process::exit(1);
    });
    (languages, iterations, variant)
}
